### ISSUE TITLE
How to use _RETR_usingReadFile/_RETR_usingCreateReadStream without an actual file 

### ISSUE THREAD

**Issue by SameerTi** [nodeftpd/nodeftpd#130]
Following is sample code

However in my case, this.createReadStream is used to perform FTP download, however there's no real file being downloaded here, i need to deal with Buffer/Streams here

Having said, that, createReadStream gets following PATH as request:
node-linux-x64-4-1-1/download/1493213659892150406dcfa7fd4a43b9f62c3d9fb4620/**50**

Last 50 actually shows size in MB, how can i work this out?
`
var ftpServer=new ftpd.FtpServer(host,options);

	ftpServer.on('client:connected',function(conn) {
	var rfs=new restrictedFs(conn);
	var username;
	logger.debug('Ftp client connected from ' + conn.socket.remoteAddress);
	conn.on('command:user',function(user,success,failure) {
		username=user;
		(user == appConfig.ftpUsername) ? success() : failure();
	});

	conn.on('command:pass',function(pass,success,failure) {
		// check the password 
		(pass == appConfig.ftpPassword) ? success(username,rfs) : failure();
	});
`
	var restrictedFs=function(conn) {
	var ftpStartTime=Date.now();
	this.unlink=function(path,callback) {
		if (callback)
			callback();
	};

	this.readdir=function(path,callback) {
		fs.readdir(path,callback);
	};

	this.mkdir=function(path,mode,callback) {
		logger.warn('Creating new directory is not allowed ' + path);
		if (callback)
			callback();
	};

	this.rmdir=function(path,callback) {
		if (callback)
			callback();
	};

	this.rename=function(oldPath,newPath,callback) {
		if (callback)
			callback();
	};

	this.stat=function(path,callback) {
		fs.stat(path,callback);
	};

	this.createWriteStream=function(path,options) {
		return fs.createWriteStream(path,options);
	};

         this.createReadStream=function(path) {       //FTP download
		return fs.createReadStream(path);
	};

	}
`

**Comment by mk-pmb** [comment]
I'm not sure which sample code you mean; however, it looks like you're trying to use a custom file system implementation stored in the `rfs` variable.
The code you posted is cumbersome to read in its current presentation. Please make it easier to read by putting <code>&grav;&grav;&grav;javascript</code> in front of it and <code>&grav;&grav;&grav;javascript</code> after it, each on a line of its own.
From what I can guess though, it looks like you might want to make all the function declarations at the bottom refer to `restrictedFs.prototype` (via an alias variable) instead of `this`.

**Comment by SameerTi** [comment]
Indeed i'm using "custom file system", stored in rfs variable
My apologies for not pasting code properly

Here i will paste it again

`
var o_DownloadBufferSize=1024 * 1024; // 1MB Buffer
var o_DownloadBuffer=new Buffer(o_DownloadBufferSize);
o_DownloadBuffer.fill("0xFF");

var restrictedFs=function(conn) {
this.createReadStream=function(path){
   
}

...
..
..

var ftpServer=new ftpd.FtpServer(host,options);

ftpServer.on('client:connected',function(conn) {
	var rfs=new restrictedFs(conn);
        var username;
        conn.on('command:user',function(user,success,failure) {
		username=user;
		(user == appConfig.ftpUsername) ? success() : failure();
	});

	conn.on('command:pass',function(pass,success,failure) {
		// check the password 
		(pass == appConfig.ftpPassword) ? success(username,rfs) : failure();
	});

}
...
...
..
`
Above is code from my module... and as you can see, using "createReadStream" for FTP download test.
Under "createReadStream" i get path as "/download/120", this 120 is in MB
This createReadStream is called from "Ftpd" module, as in following manner:
`
FtpConnection.prototype._RETR_usingCreateReadStream = function(commandArg, filename) {
...
..

  afterOk(function() {
    self._whenDataReady(function(pasvconn) {
      **var rs = self.fs.createReadStream(filename);**
      var readLength = 0;
...
..

`
My question is if you can provide me, sample code to be used under createReadStream, which is best in terms of performance, to send 120 MB in chunk as stream in parallel to process for ftpd module

Thanks in Advance

**Comment by mk-pmb** [comment]
I think the only aspect related to ftpd is the line where you pass `rfs`:

```javascript
    (pass == appConfig.ftpPassword) ? success(username,rfs) : failure();
```

That looks correct, so the remainder of the problem is implementing a `createReadStream` function on `rfs`.

The performance-relevant parts of [lib/FtpConnection.js](https://github.com/nodeftpd/nodeftpd/blob/master/lib/FtpConnection.js) are the `.pause()` and `.pipe()` calls on the readable stream obtained from `rfs.createReadStream()`, so you should make sure your stream supports the usual back pressure features. How to do that is a general nodejs streams topic, thus out of scope for this project's issue tracker. Stream questions come up quite often on the #node.js IRC channel though, so you could ask there, or in the [general node.js help repo](https://github.com/nodejs/help).
When your stream supports the standard back pressure mechanisms, in theory, chunk size should adapt automatically and dynamically based on network saturation. In theory, any attempt to enforce custom chunk sizes will waste resources for the additional checks and timers.

**Comment by SameerTi** [comment]
Thanks for pointing me to another thread, i have created ticket for the same:
https://github.com/nodejs/help/issues/611

However, there's something that i have done, but like to get it reviewed by you, here it goes!

createReadStream is called from "ftpd" module as below:
```
var rs = self.fs.createReadStream(filename);
```
modify it to:
```
var rs = self.fs.createReadStream(filename,pasvconn);
```

Above is change as far as ftpd module is concerned, and in my module...here's what i plan to do:
```
var o_DownloadBufferSize=1024 * 1024; // 1MB Buffer
var o_DownloadBuffer=new Buffer(o_DownloadBufferSize);
o_DownloadBuffer.fill("0xFF");

for (var nBytesSentMB=1;nBytesSentMB<=120;nBytesSentMB+= 1) {   //120 here  is in MB i.e. 120 MB
		pasvconn.write(o_DownloadBuffer);
}
```

So, as you see, writing 1 MB buffer directly to pasvconn and completely eliminating any Buffer creating from picture

Please let me know, what do you think about this approach!

**Comment by mk-pmb** [comment]
Your idea could save some very few function calls per write, at a higher maintainance cost.
It deviates from the node filesystem interface, so you might get compatibility conflicts in the future if the interface is extended.
You also opt out of all workarounds that future versions of ftpd might implement to deal with abnormal FTP clients, or shield your code from unexpected `pasvconn` behavior. The most obvious warning flag is that your code doesn't check the `.write()`'s return value, but there may be other edge cases you'd want to cover. I remember when I last wrote a node server that handled TCP connections directly, I had to deal with lots of exceptions and race conditions. I'd rather rely on `.pipe()`, it's not that expensive performace-wise.
